docs/tutorial/hyperfoil.adoc

```diff
@@ -1,129 +1,156 @@
-qDup isolates scripts by assigning them to roles that are running on different `hosts`. To isolate `test-endoing` from `getting-started` we need to introduce a second role. Let us call it `hyperfoil`. The `roles` section will now look like the following:
+The previous tutorial created a role to run all of ours scripts. This works but a proper performance test will normally isolate the load generator (`hyperfoil`) from the process we are testing (`quarkus dev`).
+qDup isolates scripts by assigning them to roles that are running on different `hosts`. To isolate `test-endpoint` from `run-getting-started` we need to introduce a second role. Let us call it `hyperfoil`. The `roles` section will now look like the following:
 
 ```yaml
-# ...
 roles:
-  hyperfoil:
+  setup:
     hosts:
-    - test #we still only have 1 ssh computer
+    - test
+    setup-scripts:
+    - setup-getting-started
     run-scripts:
-    - test-endpoint
-  example:
+    - run-getting-started
+  hyperfoil:
     hosts:
     - test
+    setup-scripts:
+    - setup-hyperfoil
     run-scripts:
-    - getting-started
+    - test-endpoint
+    cleanup-scripts:
+    - remove-hyperfoil
 ```
 NOTE: We are only running an ssh server on one computer so we still use the `test` host for both roles but now we could split them if needed.
 
-=== Hyperfoil
-We named the new role `hyperfoil` but right now it is running `curl`. Open up the link:https://hyperfoil.io/quickstart/quickstart1.html[hyperfoil quickstart] ad we are going to follow it's steps in a terminal and add them to the `test-endpoint` script.
+=== Hyperfoil role
+We named the new role `hyperfoil`.
 
-NOTE: we test commands in a terminal while writing our qDup scripts to check their expected output and look for the common challenges that come with automating shell commands.
-
-The first command is to `wget` the hyperfoil release but we are going to start with `cd /tmp` like we did in `getting-started` so that we don't add files to the current home directory.
+==== setup-hyperfoil
+It will download, install Hyperfoil locally and configure the `single-request.hf.yaml` file to use `http://localhost:8080` instead of `http://hyperfoil.io`
 
 ```yaml
-#...
-  test-endpoint:
-  - sh: cd /tmp
-  - sh: > #the carrot is yaml for combine the lines separated by spaces and helps with long commands
-      wget https://github.com/Hyperfoil/Hyperfoil/releases/download/release-0.22/hyperfoil-0.22.zip
-      && unzip hyperfoil-0.22.zip
-      && cd hyperfoil-0.2
-  - wait-for: ready
-    then:
-    - sh: curl localhost:8080/hello #replace localhost if you used a different ssh computer name
-    - signal: done #tells qDup that the "testing" is done
+- sh: "sed -i 's|http://hyperfoil.io|http://localhost:8080|g' examples/single-request.hf.yaml"
 ```
 
-The next command is to run `> bin/cli.sh`. Test that in terminal and notice how it starts an interactive shell and changes the prompt.
+==== test-endpoint
```

Comment:
What is the benefit of breaking this section into multiple sub-sections with the `====` decorator? I do not understand how breaking each script into a different heading section helps the readability. Was it difficult to read in the previous format?

Reply:
> What is the benefit of breaking this section into multiple sub-sections with the ==== decorator?

It is clear that `==== test-endpoint` belongs to `=== Hyperfoil role`

> Was it difficult to read in the previous format?

It wasn't.

Here is an example:

![image](https://github.com/user-attachments/assets/f7b303f9-9e6f-4759-b5e6-60fca4137897)